activemq-broker/src/main/java/org/apache/activemq/replica/ReplicaSequenceStorage.java

```diff
@@ -0,0 +1,149 @@
+                                  ReplicaInternalMessageProducer replicaInternalMessageProducer, String sequenceName) {
+        this.broker = requireNonNull(broker);
+        this.connectionContext = connectionContext;
+        this.replicaInternalMessageProducer = replicaInternalMessageProducer;
+        this.sequenceName = requireNonNull(sequenceName);
+        this.queueProvider = queueProvider;
+
+        replicationProducerId.setConnectionId(new IdGenerator().generateId());
+    }
+
+    public String initialize() throws Exception {
+        sequenceQueue = broker.getDestinations(queueProvider.getSequenceQueue()).stream().findFirst()
+            .map(DestinationExtractor::extractQueue).orElseThrow();
+
+        String selector = String.format("%s LIKE '%s'", SEQUENCE_NAME_PROPERTY, sequenceName);
+
+        ConnectionId connectionId = new ConnectionId(new IdGenerator("ReplicationPlugin.ReplicaSequenceStorage").generateId());
+        SessionId sessionId = new SessionId(connectionId, new LongSequenceGenerator().getNextSequenceId());
+        ConsumerId consumerId = new ConsumerId(sessionId, new LongSequenceGenerator().getNextSequenceId());
+        ConsumerInfo consumerInfo = new ConsumerInfo();
+        consumerInfo.setConsumerId(consumerId);
+        consumerInfo.setPrefetchSize(10);
+        consumerInfo.setDestination(queueProvider.getSequenceQueue());
+        consumerInfo.setSelector(selector);
+        subscription = (PrefetchSubscription) broker.addConsumer(connectionContext, consumerInfo);
+
+        List<ActiveMQTextMessage> allMessages = new ArrayList<>();
+        for (MessageId messageId : sequenceQueue.getAllMessageIds()) {
+            ActiveMQTextMessage message = getMessageByMessageId(messageId);
+            if (message.getStringProperty(SEQUENCE_NAME_PROPERTY).equals(sequenceName)) {
+                allMessages.add(message);
+            }
+        }
+
+        if (allMessages.size() == 0) {
+            return null;
+        }
+
+        if (allMessages.size() > 1) {
+            for (int i = 0; i < allMessages.size() - 1; i++) {
+                sequenceQueue.removeMessage(allMessages.get(i).getMessageId().toString());
+            }
+        }
+
+        return allMessages.get(0).getText();
+    }
+
+    public void enqueue(TransactionId tid, String message) throws Exception {
+        // before enqueue message, we acknowledge all messages currently in queue.
+        acknowledgeAll(tid);
+
+        send(tid, message);
+    }
+
+    private void acknowledgeAll(TransactionId tid) throws Exception {
+        List<MessageReference> dispatched = subscription.getDispatched();
+        ConsumerBrokerExchange consumerExchange = new ConsumerBrokerExchange();
+        consumerExchange.setConnectionContext(connectionContext);
+        consumerExchange.setSubscription(subscription);
+
+        for(MessageReference messageReference: dispatched) {
```

Comment:
Is it any cheaper to send 1 ack for the whole range? 

ack.setFirstMessage(...)
ack.setLastMessage(...)
ack.setMessageCount(...)